File: OJ/leetcode/cpp/39_combination_sum.cpp
#include <iostream>
#include <vector>
#include <numeric>
using namespace std;

void dfs(int start, vector<int>& stack, vector<int>& candidates, int target, vector<vector<int>>& ans) {
    if (target <= 0) {
        if (target == 0) {
            ans.push_back(stack);
        }
        return;
    }
    for (int i = start; i < candidates.size(); ++i) {
        stack.push_back(candidates[i]);
        dfs(i, stack, candidates,  target - candidates[i], ans);
        stack.pop_back();
    }
}

vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<vector<int>> ans;
    vector<int> stack;
    dfs(0, stack, candidates, target, ans);
    return ans;
}

int main() {
    vector<int> candidates = {2, 3, 6, 7};
    vector<vector<int>> ans;
    ans = combinationSum(candidates, 7);
    for (vector<int> vec : ans) {
        for (int v : vec) {
            cout << v << " ";
        }
        cout << "\n";
    }
}

File: competitive_programming/leetcode/python/39_combination_sum.py
def bfs(candidates, target):
  n = len(candidates)
  ans = []
  tree = [[i] for i in range(n)]
  count = target // min(candidates) + 1
  
  while count > 0:
    tmp = []
    for arr in tree:
      sum_ = sum([candidates[i] for i in arr])
      if sum_ >= target:
        if sum_ == target:
          ans.append([candidates[i] for i in arr])
        continue
      for i in range(arr[-1], n):
        tmp.append(arr + [i])
    tree = tmp
    count -= 1

  return ans

def dfsSolution(candidates, target):
  def dfs(candidates, target, start, stack, ans):
    if target == 0:
      ans.append(stack[:])
      return
    
    for i in range(start, len(candidates)):
      if candidates[i] > target:
          return 
      stack.append(candidates[i])
      dfs(candidates, target - candidates[i], i, stack, ans)
      stack.pop()

  candidates.sort()
  ans = []
  dfs(candidates, target, 0, [], ans)
  return ans

print(dfsSolution([2, 3, 6, 7], 7))
print(dfsSolution([2, 3, 5], 8))
print(dfsSolution([2], 1))
print(dfsSolution([1], 1))
print(dfsSolution([1], 2))
print(dfsSolution([10, 1, 2, 7, 6, 1, 5], 8))
print(dfsSolution(
    [1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1, 
     1, 1, 1, 1, 1], 27
    )
)